CloudSQLReplicator/google/cloud/pontem/sql/replicator/util/gcp_api_util.py

```diff
@@ -13,20 +13,133 @@
 # limitations under the License.
 
 """GCP API utility functions."""
+import httplib2
 import logging
+import uuid
+
 import google.auth
 from google.cloud import storage
+import google_auth_httplib2
+from googleapiclient import discovery
 
 from google.cloud.pontem.sql import replicator
 
-
+DEFAULT_TIER = 'db-n1-standard-2'
 
 def _get_user_agent():
   """Returns user agent based on packagage info."""
 
   user_agent = 'pontem,{}/{} (gzip)'.format(replicator.__package_name__, replicator.__version__)
   return user_agent
 
+def _get_user_agent_header():
+  """Returns custom User-Agent header."""
+
+  user_agent = _get_user_agent()
+  headers = {'User-Agent': user_agent}
+  return headers
+
+# SQL Admin
+def build_sql_admin_service(credentials=None):
+  """Authorizes and sets custom CloudSQL Replicator user agent.
+
+    Args:
+      project (str): Project ID
+      credentials (google.auth.Credentials): Credentials to authorize client
+
+    Returns:
+      Authorized sqladmin service with CloudSQL Replicator user agent.
+  """
+  headers = _get_user_agent_header()
+  httplib2.Http.request.__func__.func_defaults = ('GET', None, headers, 5, None)
+  default_credentials, _ = google.auth.default()
+  authorized_http = google_auth_httplib2.AuthorizedHttp(
+      credentials or default_credentials
+  )
+
+  service = discovery.build(
```

Comment:
Can this fail if there are insufficient permissions? How do we report the error (stackdump?)